keps/sig-node/20191127-cri-containerstats-enhancements.md

```diff
@@ -0,0 +1,201 @@
+---
+title: Enhance ContainerStats message in CRI-API
+authors:
+  - "@fuweid"
+owning-sig: sig-node
+participating-sigs:
+  - sig-node
+  - sig-instrumentation
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-27
+last-updated: 2019-12-09
+status: provisional
+---
+
+# Enhance ContainerStats message in CRI-API
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This enhancement is to extend `CpuUsage` and `MemoryUsage` fields in
+`ContainerStats`. And also add new `ProcessUsage`/`NetworkUsage` field in
+`ContainerStats`. The Container Runtime Engine will cover more metrics about
+managed container and let the metric collectors focus on node-level metrics or
+monitoring metrics.
+
+## Motivation
+
+Since more and more container runtimes shows up, like `kata-container`/
+`gVisor`/ `Firecracker`/ `Windows Containers`, the way to collect metrics from
+these containers is quite different from runC. For the third-party metric
+collectors, they have to maintain the knowledge about how to collect metrics
+from each kind of container runtimes. And more, kubelet codebase focus on
+core-metrics export (defined in [Kubernetes monitoring architecture#Terminology](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#terminology)),
+and there is no powerful standard interface for container level metrics.
+It will bring burden to end-users who wants to build powerful monitor pipeline
+(defined in [Kubernetes monitoring architecture#Executive Summary](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) for different container runtimes.
+
+Based on this case, we can extend Container Runtime Interface(CRI) to define
+the container level metrics. Each container runtime engine, which implemented
+CRI, knows lifecycle and metrics about each managed container better than any
+external metric collector. For the third-party metric collectors, they just
+need to follow the `ContainerStats` API defined by CRI. With standard CRI,
```

Comment:
I like the idea to extend CRI api with metrics and already discussed it with @brancz.
Main concern was to make sure that this API is designed with close collaboration of runtime developers to avoid reimplementing Summary API. There are projects where CRI is implemented in unexpected ways (VMs, separate nodes etc). It would be useful to have some analysis of more popular project to get some better understanding what metric means for different runtime.

I see couple of cons about exposing CRI api to metrics collectors directly.
* Authorization - Do we need to develop full authorization mechanism to just expose metrics. Kubelet has already solved this problem for us, do we need separate solution
* Cannot expose outside node - This API would not be a good fit to replace Summary API as having CRI exposed outside node is too big attack surface. 
* Proto API for metrics - It makes integrations with metrics collectors harder. We should use one mechanism for instrumenting metrics in k8s using framework that we created. 

Having two APIs (one for CRI, second for kubelet) has some downsides like you listed, but it creates clear responsibility separation separation. Runtime can focus on measuring resources and kubelet can handle Authorization and Metric Lifecycle. Overlap can be good as both API can have different requirements for deprecation and version drift.

Reply:
Thanks for the summary! 

> Main concern was to make sure that this API is designed with close collaboration of runtime developers to avoid reimplementing Summary API. 

IIUC, we can make good summary for common container metrics. CGroups has good summary about resources, especially in v2 version. Also `Windows stats are slightly different` from @jterry75 . CRI-API is standard so that `ContainerStats` must have common metrics terms. For the special runtime and special metrics, I think it makes senses to use special agent to collect metrics.

> Authorization & Cannot expose outside node

Thanks for mention about Auth! Original thinking is that metrics collector agent is aim to expose container metrics outside node without auth. Also I just realized that CRI didn't handle auth right now. Anyone can mount socket into container and use `crictl` to do anything. At least, collecting metrics directly from CRI doesn't make it worse. :) 

Anyway, I agree with you about `Runtime can focus on measuring resources and kubelet can handle Authorization and Metric Lifecycle.` Mentioned in https://github.com/kubernetes/kubernetes/issues/68522, the goal is to provide plug-in functionality to support out-of-tree source metrics. And CRI ContainerStats is one of source. Maybe @dashpole can provide more details about this :)

Reply:
That proposal is specifically for device metrics.  As a vendor-neutral open-source project, we should not have vendor-specific device monitoring code in our components if at all possible.  The same cannot be said for "generic" container stats, which are fine to support.

I completely agree with @serathius' point about auth.  Pods must be privileged in order to mount the CRI socket with a hostpath volume.  This proposal would mean that all metrics collection agents must be privileged on the node.  Beyond that, it isn't possible to allow access to metrics without granting root on the node (since you can run arbitrarily privileged pods through the CRI).

I also agree with the point about using a non-standard (proto) API.  The current problem with the summary API is that collectors have to write additional clients for collecting metrics.  While grpc is easier (since you can use generated clients), you still have to write a new collector to get these metrics.